sdk/include/ndk/haltypes.h

```diff
@@ -181,22 +1729,336 @@ typedef struct _HAL_PRIVATE_DISPATCH
+    pHalTimerQueryAuxiliaryCounterFrequency HalTimerQueryAuxiliaryCounterFrequency;
+    pHalConnectThermalInterrupt HalConnectThermalInterrupt;
+    pHalIsEFIRuntimeActive HalIsEFIRuntimeActive;
 #endif
 } HAL_PRIVATE_DISPATCH, *PHAL_PRIVATE_DISPATCH;
 
+typedef struct _IOMMU_DEVICE_PATH
+{
+    GUID BusTypeGuid;
+    ULONG UniqueIdLength;
+    ULONG PathLength;
+    PVOID UniqueId;
+    PVOID Path;
+} IOMMU_DEVICE_PATH, *PIOMMU_DEVICE_PATH;
+
+typedef union _IOMMU_SVM_CAPABILITIES
+{
+    struct {
+        ULONG AtsCapability:1;
+        ULONG PriCapability:1;
+        ULONG PasidCapability:1;
+        struct {
+            ULONG PasidMaxWidth:5;
+            ULONG PasidExePerm:1;
+            ULONG PasidPrivMode:1;
+            ULONG AtsPageAlignedRequest:1;
+            ULONG AtsGlobalInvalidate:1;
+            ULONG AtsInvalidateQueueDepth:5;
+        } CapReg;
+        ULONG Rsvd:15;
+    } DUMMYSTRUCTNAME;
+    ULONG AsULONG;
+} IOMMU_SVM_CAPABILITIES, *PIOMMU_SVM_CAPABILITIES;
+
+//
+// HAL IOMMU function Types
+//
+typedef
+BOOLEAN
+(NTAPI *pHalIommuSupportEnabled)(
+    VOID
+);
+
+typedef
+_IRQL_requires_max_(PASSIVE_LEVEL)
+NTSTATUS
+(NTAPI *pHalIommuGetConfiguration)(
+    _In_ ULONG Domain,
+    _Out_ PULONG PageRequestQueues,
+    _Out_ PULONG MaximumAsids,
+    _Out_ PVOID *SystemContext
+);
+
+typedef
+_IRQL_requires_max_(PASSIVE_LEVEL)
+NTSTATUS
+(NTAPI *pHalIommuGetLibraryContext)(
+    _In_ ULONG Pasid,
+    _In_ ULONG Domain,
+    _Out_ PVOID *LibraryContext
+);
+
+typedef
+_IRQL_requires_max_(PASSIVE_LEVEL)
+NTSTATUS
+(NTAPI *pHalIommuMapDevice)(
+    _In_ PVOID LibraryContext,
+    _In_ PIOMMU_DEVICE_PATH DevicePath,
+    _In_ PIOMMU_SVM_CAPABILITIES DeviceCapabilities,
+    _Out_ PVOID *DeviceHandle
+);
+
+typedef
+_IRQL_requires_max_(PASSIVE_LEVEL)
+NTSTATUS
+(NTAPI *pHalIommuEnableDevicePasid)(
+    _In_ PVOID LibraryContext,
+    _In_ PVOID DeviceHandle
+);
+
+typedef
+_IRQL_requires_max_(PASSIVE_LEVEL)
+NTSTATUS
+(NTAPI *pHalIommuSetAddressSpace)(
+    _In_ PVOID LibraryContext,
+    _In_ ULONG_PTR DirectoryBase
+);
+
+typedef
+_IRQL_requires_max_(PASSIVE_LEVEL)
+NTSTATUS
+(NTAPI *pHalIommuDisableDevicePasid)(
+    _In_ PVOID LibraryContext,
+    _In_ PVOID DeviceHandle
+);
+
+typedef
+_IRQL_requires_max_(PASSIVE_LEVEL)
+NTSTATUS
+(NTAPI *pHalIommuUnmapDevice)(
+    _In_ PVOID SystemContext,
+    _In_ PVOID DeviceHandle
+);
+
+typedef
+_IRQL_requires_max_(PASSIVE_LEVEL)
+NTSTATUS
+(NTAPI *pHalIommuFreeLibraryContext)(
+    _In_ PVOID LibraryContext
+);
+
+typedef
+_IRQL_requires_max_(HIGH_LEVEL)
+VOID
+(NTAPI *pHalIommuFlushTb)(
+    _In_ PVOID LibraryContext,
+    _In_ ULONG Number,
+    _In_count_ (Number) KTB_FLUSH_VA Virtual[]
+);
+
+typedef
+_IRQL_requires_max_(HIGH_LEVEL)
+VOID
+(NTAPI *pHalIommuFlushAllPasid)(
+    _In_ PVOID LibraryContext,
+    _In_ ULONG Number,
+    _In_count_ (Number) KTB_FLUSH_VA VA[]
```

Comment:
```suggestion
    _In_count_(Number) KTB_FLUSH_VA VA[]
```
and `VA` --> `Virtual` ?

Reply:
> `KTB_FLUSH_VA VA[]`

Why not using `KTB_FLUSH_VA* pVA` instead? (maybe this question can be understood as: did you copy the definition from the headers of some other project?)

Reply:
Yes, it is defined like this in WDK. Although non-array pointer definition seems to make more sense nonetheless... Would it preferable, here and in other occurrences?

Reply:
ok let's keep the array notation if this is what MS uses officially. This should give the same signature ultimately :thinking: